### ISSUE TITLE
unexpected behavior when use union type

### ISSUE THREAD

**Issue by MaoScut** [Microsoft/TypeScript#28860]
**TypeScript Version:**  3.1.4

**Search Terms:**
union type
**Code**

```ts
export type basetype = {
  id: string;
}
export type supertype1 = basetype & {}
export type supertype2 = basetype & {
  key1: any;
  key2: any;
}

function myfun(p: supertype2 | supertype1) {
  return 
}

myfun( {
  id: '',
  key1: '',
})
```

**Expected behavior:**
 {
  id: '',
  key1: '',
} is not supertype1 or supertype2, when it is used to be a parameter of myfun, typescript should report an error.
**Actual behavior:**
there is no error.

**Comment by j-oliveras** [comment]
The `myfun` parameter is assignable to `supertype1` (that is basically the same as `basetype`). Remember that the type are [checked structurally](https://github.com/Microsoft/TypeScript/wiki/FAQ#what-is-structural-typing).

Maybe that you expect is a exact type. See discussions about that: #202, #12936 and PR #28749.

**Comment by ahejlsberg** [comment]
Write it like this instead:

```ts
export type basetype = { id: string }

export type supertype1 = basetype & { key1?: undefined, key2?: undefined }
export type supertype2 = basetype & { key1: any, key2: any; }

function myfun(p: supertype2 | supertype1) {}

myfun({ id: '' });  // Ok
myfun({ id: '', key1: '' });  // Error
myfun({ id: '', key1: '', key2: '' });  // Ok
```

**Comment by MaoScut** [comment]
> The `myfun` parameter is assignable to `supertype1` (that is basically the same as `basetype`). Remember that the type are [checked structurally](https://github.com/Microsoft/TypeScript/wiki/FAQ#what-is-structural-typing).
> 
> Maybe that you expect is a exact type. See discussions about that: #202, #12936 and PR #28749.

I know structural typing, so I pass an object literal instead of a variable to myfun. And Exact type is what I need. Thank you.

**Comment by MaoScut** [comment]
> Write it like this instead:
> 
> ```ts
> export type basetype = { id: string }
> 
> export type supertype1 = basetype & { key1?: undefined, key2?: undefined }
> export type supertype2 = basetype & { key1: any, key2: any; }
> 
> function myfun(p: supertype2 | supertype1) {}
> 
> myfun({ id: '' });  // Ok
> myfun({ id: '', key1: '' });  // Error
> myfun({ id: '', key1: '', key2: '' });  // Ok
> ```

Nice. It work for me. Thank you.
`interface { key1?: undefined} `
describes key1 is a forbidden property. it's better to add this syntax to TypeScript documents.(handbook --> interface --> Optional Properties)
